mne/coreg.py

```diff
@@ -1242,3 +1245,572 @@ def _scale_xfm(subject_to, xfm_fname, mri_name, subject_from, scale,
                 F_mri_ras, 'ras', 'ras'),
             F_ras_mni, 'ras', 'mni_tal')
     _write_fs_xfm(fname_to, T_ras_mni['trans'], kind)
+
+
+class Coregistration(object):
+    """Class for MRI<->head coregistration.
+
+    Parameters
+    ----------
+    info : instance of Info
+        The measurement info.
+    subject : str
+        Name of the subject the data are defined for.
+    subjects_dir : path-like
+        Path to MRI subjects directory.
+
+    Notes
+    -----
+    Internal computation quantities parameters are in units of (for X/Y/Z):
+    - rotation are in radians
+    - translation are in m
+    - scale are in scale proportion
+    """
+
+    def __init__(self, info, subject, subjects_dir):
+        self._rot_trans = None
+        self._default_parameters = \
+            np.array([0., 0., 0., 0., 0., 0., 1., 1., 1.])
+        self._n_scale_params = (0, 1, 3)
+
+        self._icp_iterations = 20
+        self._icp_angle = 0.2
+        self._icp_distance = 0.2
+        self._icp_scale = 0.2
+        self._icp_fid_matches = ('nearest', 'matched')
+        self._icp_fid_match = self._icp_fid_matches[0]
+        self._lpa_weight = 1.
+        self._nasion_weight = 10.
+        self._rpa_weight = 1.
+        self._hsp_weight = 1.
+        self._eeg_weight = 1.
+        self._hpi_weight = 1.
+
+        self._dig = _DigSource(info)
+        self._mri = _MRIHeadWithFiducialsModel(subjects_dir=subjects_dir,
+                                               subject=subject)
+        self.reset()
+        self._nearest_calc = _DistanceQuery(
+            self._processed_high_res_mri_points * self._scale)
+
+    def _update_params(self, rot=None, tra=None, sca=None):
+        rot_changed = False
+        if self._rot_trans is None or rotation is not None:
+            rot_changed = True
+            self._last_rotation = self._rotation
+            self._rotation = rot
+            self._rot_trans = rotation(*rot).T
+        tra_changed = False
+        if rot_changed or tra is not None:
+            tra_changed = True
+            self._last_translation = self._translation
+            self._translation = tra
+            self._head_mri_t = self._rot_trans.copy()
+            self._head_mri_t[:3, 3] = \
+                -np.dot(self._head_mri_t[:3, :3], tra)
+            self._transformed_dig_hpi = \
+                apply_trans(self._head_mri_t, self._dig.hpi_points)
+            self._transformed_dig_eeg = \
+                apply_trans(self._head_mri_t, self._dig.eeg_points)
+            self._transformed_dig_extra = \
+                apply_trans(self._head_mri_t, self._dig.extra_points)
+            self._transformed_orig_dig_extra = \
+                apply_trans(self._head_mri_t, self._dig._extra_points)
+            self._nearest_transformed_high_res_mri_idx_orig_hsp = \
+                self._nearest_calc.query(
+                    apply_trans(self._head_mri_t, self._dig._extra_points))[1]
```

Comment:
The order helps us here, as we can do things like this:
```suggestion
                self._nearest_calc.query(self._transformed_orig_dig_extra)[1]
```